### PR TITLE
HPCC-15618 Systemd Modifications for 7.0 Beta

### PR DESCRIPTION
**Pull request by Michael-Gardner**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by Michael-Gardner** [comment]
@mckellyln This PR also will cover the Limit issue that you raised with https://track.hpccsystems.com/browse/HPCC-19030.  The templates modify the RTPRIO, memlock, nproc, nofile, and core limits.

File: initfiles/bash/etc/systemd/system/hpcc-environment-monitor.install.in [hunk 1]
```diff
@@ -0,0 +1,4 @@
+if which systemctl 1>/dev/null 2>&1; then
```
Comment (mckellyln) [inline initfiles/bash/etc/systemd/system/hpcc-environment-monitor.install.in]: Is this supposed to run if systemctl is found is is NOT found ?

File: initfiles/sbin/add_conf_settings.sh.in [hunk 2]
```diff
@@ -31,15 +31,19 @@ if grep "\\#includedir /etc/sudoers.d" /etc/sudoers > /dev/null 2>&1; then
...
+${USER_NAME} ALL = NOPASSWD: HPCC_DAFILESRV, HPCC_HPCCINIT, HPCC_SYSD_THORSLAVE_START, HPCC_SYSD_THORSLAVE_STOP
 Defaults:${USER_NAME} !requiretty
 EOF
   chmod 0440 /etc/sudoers.d/${USER_NAME}
 else
   alter_file /etc/sudoers "^Cmnd_Alias HPCC_|^${USER_NAME} ALL =|^Defaults\:${USER_NAME} !requiretty" <<-%EOF
 Cmnd_Alias HPCC_DAFILESRV = /etc/init.d/dafilesrv
 Cmnd_Alias HPCC_HPCCINIT = /etc/init.d/hpcc-init
-${USER_NAME} ALL = NOPASSWD: HPCC_DAFILESRV, HPCC_HPCCINIT
+Cmnd_Alias HPCC_SYSD_THORSLAVE_START = /bin/systemctl start thorslave*
```
Comment (mckellyln) [inline initfiles/sbin/add_conf_settings.sh.in]: Why use which systemctl above but /bin/systemctl here ?

File: thorlcr/slave/thslavemain.cpp [hunk 3]
```diff
@@ -296,6 +296,15 @@ void setSlaveAffinity(unsigned processOnNode)
...
 {
+    for (unsigned i=0;i<(unsigned)argc;i++) {
+        if (strcmp("--daemon",argv[i])==0 || strcmp("-d",argv[i])==0) {
+            if (daemon(1,0) || write_pidfile(argv[++i])) {
+                perror("Failed to daemonize");
+                return EXIT_FAILURE;
+            }
+            break;
+        }
+    }
```
Comment (mckellyln) [inline thorlcr/slave/thslavemain.cpp]: Does the ssh (frunssh) from init_thor work ok to start all remote slaves and know they are ok if they now daemonize themselves immediately ?

**Comment by mckellyln** [comment]
@Michael-Gardner a few questions inline.

**Comment by richardkchapman** [comment]
Typo in commit message

**Comment by Michael-Gardner** [comment]
@richardkchapman rebased, fixed merge issues, and squashed commits.

File: initfiles/bash/etc/systemd/system/README.md [hunk 4]
```diff
@@ -0,0 +1,21 @@
+Systemd scripts for HPCC-Systems Platform
+=========================================
+
+  The primary tool for starting and stopping the entire platform on a target machine using systemd is the `hpccsystems-platform.target`.  This target script is entirely dependent on the `environment.xml` file located on the server, and gets generated using the `generate-hpccsystems-target.sh` script that is located in `/opt/HPCCSystems/sbin` on standard installations.  This target script will need to be generated any time you modify the environment.xml on the local machine.  This can be done manually by an administrator, or by enabling the `hpcc-environment-monitor.path` systemd script which we provide.
```
Comment (richardkchapman) [inline initfiles/bash/etc/systemd/system/README.md]: These are long lines. Should wrap at 80 cols really...

**Comment by Michael-Gardner** [comment]
I was told not to daemonize the thor slaves.

The older sysV scripts still work, and can still be called directly by administrators/tools to get the older behavior.  After a core, systemd will restart the process with the original arguments that it was passed, and keep attempting until successful.  Not certain on Roxie reporting, I'll look into that.

**Comment by Michael-Gardner** [comment]
Also I have a systemd compatible hpcc-run.sh script that should make life a bit easier for people.  I'll push the PR for it in a few minutes.

File: initfiles/bash/etc/systemd/system/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+set(dafilesrv "dafilesrv" "-L ${LOG_PATH} -I %i -D" "")
+set(dali "daserver" "--daemon %i" "dafilesrv.service")
+set(dfuserver "dfuserver" "--daemon %i" "dafilesrv.service")
+set(eclagent "agentexec" "--daemon %i" "dafilesrv.service")
+set(eclccserver "eclccserver" "--daemon %i" "dafilesrv.service")
+set(eclscheduler "eclscheduler" "--daemon %i" "dafilesrv.service")
+set(esp "esp" "--daemon %i" "dafilesrv.service")
+set(roxie "roxie" "--topology=RoxieTopology.xml --logfile --restarts=2 --stdlog=0 --daemon %i" "dafilesrv.service")
+set(sasha "saserver" "--daemon %i" "dafilesrv.service" "Metadata in-memory store.")
```
Comment (jakesmith) [inline initfiles/bash/etc/systemd/system/CMakeLists.txt]: Not sure why the last value ("Metadata in-memory store.") is here?
Comment relates to Dali not Sasha.

File: initfiles/bash/etc/systemd/system/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+set(backupnode "backup.sh" "" "")
+
+foreach(component ${componentList})
+    list(GET "${component}" 0 processName)
+    list(GET "${component}" 1 processArgs)
+    list(GET "${component}" 2 dependencies)
+    if("${component}" STREQUAL "thor" OR "${component}" STREQUAL "thorslave" OR "${component}" STREQUAL "hpcc-conf" OR "${component}" STREQUAL "thor-conf" OR "${component}" STREQUAL "backupnode")
+        configure_file(${component}@instance.service.in ${component}@.service @ONLY)
+    else()
+        configure_file(component@instance.service.in ${component}@.service @ONLY)
```
Comment (jakesmith) [inline initfiles/bash/etc/systemd/system/CMakeLists.txt]: Could you comment on this if / else condition and the effect the 2 different configure_file lines have ?

File: initfiles/bash/etc/systemd/system/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    configure_file(service-unit.install.in ${component}-service.install @ONLY)
+    configure_file(service-unit.uninstall.in ${component}-service.uninstall @ONLY)
+    install(FILES ${CMAKE_CURRENT_BINARY_DIR}/${component}-service.install DESTINATION etc/init.d/install COMPONENT Systemd)
+    install(FILES ${CMAKE_CURRENT_BINARY_DIR}/${component}-service.uninstall DESTINATION etc/init.d/uninstall COMPONENT Systemd)
+    install(FILES ${CMAKE_CURRENT_BINARY_DIR}/${component}@.service DESTINATION etc/systemd/system COMPONENT Systemd)
+endforeach()
+
+## Generate and setup standalone service files
+set(standaloneList
+	dropzone
```
Comment (jakesmith) [inline initfiles/bash/etc/systemd/system/CMakeLists.txt]: trivial: out of kilter indentation

**Comment by jakesmith** [comment]
@Michael-Gardner - some comments/questions above.

@xwang2713 - please also review.

File: initfiles/bash/etc/systemd/system/CMakeLists.txt (hunk 6 above)
Comment (xwang2713) [reply initfiles/bash/etc/systemd/system/CMakeLists.txt]: component@instance.service.in is generic systemd service file. ${component}@instance.service.in is particular service file listed in if condition.

**Review by xwang2713 — COMMENTED** [review]
We never officially support Ubuntu 17.04 which will be expired in July 2018. I did check master branch already has the same setting for Ubuntu 18.04.

File: initfiles/bash/etc/systemd/system/CMakeLists.txt (hunk 5 above)
Comment (Michael-Gardner) [reply initfiles/bash/etc/systemd/system/CMakeLists.txt]: Needs to be removed, missed it in my last pass I"ll get it now.

File: initfiles/bash/etc/systemd/system/CMakeLists.txt (hunk 6 above)
Comment (Michael-Gardner) [reply initfiles/bash/etc/systemd/system/CMakeLists.txt]: Ming is correct.  There is a default service file template that I've created, and then some special service files for things like thor/thorslaves that have extra steps.

File: esp/platform/espp.cpp [hunk 8]
```diff
@@ -231,6 +232,7 @@ static void usage()
     puts("  esp [options]");
     puts("Options:");
     puts("  -?/-h: show this help page");
+    puts("  --daemon <instance>: start in background as daemon");
```
[the remark this answers is not in the document]
Comment (Michael-Gardner) [reply esp/platform/espp.cpp]: It is used in naming the pidfile on the file system.  <instance> in this case could be myesp, and the pidfile will get generated using that name.

File: dali/server/daserver.cpp [hunk 9]
```diff
@@ -139,18 +170,11 @@ int main(int argc, char* argv[])
...
-            if (myrank>=(unsigned)(argc-2)) {
-                printf("incorrect rank\n");
-                return 0;
-            }
-        }
+//TODO:
+//            if (myrank>=(unsigned)(argc-2)) {
+//                printf("incorrect rank\n");
+//                return 0;
+//            }
```
[the remark this answers is not in the document]
Comment (Michael-Gardner) [reply dali/server/daserver.cpp]: Will do

File: dali/dfu/dfuserver.cpp [hunk 10]
```diff
@@ -99,6 +99,15 @@ IPropertyTree *readOldIni()
 
 int main(int argc, const char *argv[])
 {
+    for (unsigned i=0;i<(unsigned)argc;i++) {
+        if (strcmp("--daemon",argv[i])==0 || strcmp("-d",argv[i])==0) {
```
[the remark this answers is not in the document]
Comment (Michael-Gardner) [reply dali/dfu/dfuserver.cpp]: I'll swap them over to using streq like you suggested.

**Comment by richardkchapman** [comment]
@jakesmith

File: esp/platform/espp.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (jakesmith) [reply esp/platform/espp.cpp]: daserver has : 
```
"--daemon|-d <instanceName>\t: run daemon as instance\n"
```

i.e. an optional name.

Should all --daemon's be same usage ?
Should all instance/instaname's be optional?
What is the name if i't optionally not provided? Is it / should it be e.g. constructed from the component+port ?

**Comment by jakesmith** [comment]
@Michael-Gardner - one follow up question from earlier reply.

**Comment by Michael-Gardner** [comment]
@jakesmith Changed the usage messages to the same.  Slight formatting differences on some just so they keep the same look as the rest of each components Usage message.

**Comment by jakesmith** [comment]
@Michael-Gardner - don't see any other issues, except typo' in PR title, if you could fix that and squash.

**Comment by Michael-Gardner** [comment]
@richardkchapman Squashed and ready for merge

**Pull request merged**